### ISSUE TITLE
clippy::large_futures does not fire when the .await comes from a macro expansion (e.g. #[tauri::command])

### ISSUE THREAD

**Issue by dIT8Zv** [rust-lang/rust-clippy#17586]
### Summary

`clippy::large_futures` does not fire when the `.await` expression is produced by a macro expansion, even though the awaited future is just as large as in the direct-call case. The only difference is whether the `.await` itself is macro-generated.

### Version

```
clippy 0.1.93 (01f6ddf758 2026-02-11)
rustc 1.93.1 (01f6ddf75 2026-02-11)
```

### Reproduction (no dependencies)

```rust
#![allow(dead_code)]

async fn yield_now() {}

async fn big() -> u8 {
    let big = [0u8; 65536];
    yield_now().await;
    big[0]
}

async fn direct_caller() {
    let _v = big().await; // ✔ fires: "large future with a size of 65538 bytes"
}

macro_rules! call_it {
    ($f:expr) => {
        $f.await
    };
}

async fn macro_caller() {
    let _v = call_it!(big()); // ✗ silently passes
}

fn main() {}
```

```
cargo clippy -- -W clippy::large_futures
```

Only `direct_caller` is flagged. Replacing the macro with a plain `.await` on the same expression makes the lint fire, so the future size itself is clearly computable in both cases — the `from_expansion()` guard on the await expression appears to suppress it. The same behavior occurs with a crate-level `#![warn(clippy::large_futures)]` instead of the command-line flag, and with `-D` instead of `-W`.

### Not covered by the existing test

The `macro_expn` case in `tests/ui/large_futures.rs` does expand a macro, but the macro produces the `async` block (the future itself) while the `.await` is hand-written outside the macro (`macro_!().await`). I verified that running that test file as-is does produce the annotated warnings on this toolchain — but the case where **the `.await` itself comes from macro expansion** (future from a clean call site, await generated inside the macro) is not exercised by the test suite.

The `from_expansion()` exclusion may well be intentional to avoid lint noise inside macros; the problem is that it also exempts the pattern where a framework's macro polls a user-provided future, which is exactly where large futures bite.

### Why this matters in practice

Frameworks that poll user futures through macro-generated code are where large futures cause real stack overflows:

- Tauri's `#[tauri::command]` on an `async fn` polls the command future through macro-generated wrapper code. We recently debugged a production crash (`0xC00000FD` stack overflow on Windows, no logs since the kernel kills the process) caused by a ~329 KB command future blowing up the invoke-dispatch stack frame. The dispatch chain held ~993 KiB of frames in total. Fixed by `Box::pin`-ing the future inside the command body — but `clippy::large_futures` never flagged it, because the await point lives inside the macro expansion.
- In debug builds Tauri itself `Box::pin`s command futures (for faster compiles), so the problem is release-only and invisible during development, which makes a working lint the natural first line of defense.

### Meta

Tried variants to rule out other causes: awaiting across an `async fn` boundary vs. an `async` block, `tokio::spawn`, `future-size-threshold` in `clippy.toml` (note: changing `clippy.toml` alone does not invalidate cargo's fingerprint cache — needs a `touch` on the sources to re-run), a full `#[tauri::command]` reproduction, and crate-level attributes vs. command-line flags. Direct awaits fire in every variant; macro-expanded awaits never do.

**Comment by dIT8Zv** [comment]
Correction to the original report: I initially described this as possibly contradicting the `macro_expn` case in `tests/ui/large_futures.rs`. On closer reading that case expands the `async` block but keeps the `.await` hand-written outside the macro, so it does not cover this scenario. Running the official test file as-is does reproduce its annotated warnings on this toolchain. The body has been updated to reflect this — this is a coverage gap around the `from_expansion()` exclusion rather than a regression.

**Comment by 0xull** [comment]
@rustbot claim

hi @dIT8Zv , i looked into the `clippy::large_futures` call site. i think i get why the `from_expansion` check is there, to avoid flagging code inside a macro the user can't edit (hence checking the span of the whole `.await` expression).

still finding my footing around contributing to rust so i'm open to help from the community. but at the moment i'm thinking changing the check to `arg.span` instead of `expr.span`. so the lint asks if the future itself is something the user actually wrote and it can act on.

one thing though, `tauri::command` is not `macro_rules!`, so no guarantee this actually fixes your crash, that comes down to whether tauri's own macro keeps the original span when it re-emits the command body. do you know if that's the case there?

i would open a pr soon enough.

**Comment by 0xull** [comment]
hi @dIT8Zv, i opened a draft PR for this, please see pr #17698. 

the one thing i couldn't confirm is your actual case as `#[tauri::command]` is a proc macro. but i tested against a proc macro that does preserve the span and it fires, so it comes down to how tauri is written. please if you still have that repro handy, running the branch against it would tell us for real.

**Comment by dIT8Zv** [comment]
Ran the branch against the real repro — **it fires.** That also answers your question from the earlier comment: tauri's `#[tauri::command]` expansion does keep the original spans.

**Setup**

- Project: the production Tauri 2 app behind this issue, checked out at the exact pre-fix commit from the incident — before any `Box::pin` mitigation landed, so the offending futures are still in their raw form.
- Toolchain: `nightly-2026-09-01` for both runs. Patched: your branch (`clippy 0.1.100 (cfcb2823ba 2026-09-07)`). Control: the official `clippy` component of that nightly (`clippy 0.1.100 (0dfb098f3a 2026-08-31)`). Identical crate and flags for both (`cargo clippy -- -W clippy::large_futures`, default threshold).

**Results**

Control: **1 warning** — a hand-written `.await` inside `install_modpack`'s body (36,208 bytes), which the patch leaves untouched as expected.

Patched: **5 warnings** — that same one, plus four command-dispatch futures. Each puts the primary span on the `#[tauri::command]` attribute in user code, with the `generate_handler![...]` registration in `main.rs` as the secondary span:

```text
warning: large future with a size of 73032 bytes
   --> src/commands/modpack.rs:1228:1
1228 |   #[tauri::command]            <-- install_modpack, the command from the crash
     |   ^^^^^^^^^^^^^^^^^ help: consider `Box::pin` on it
     |
    ::: src/main.rs:227:25
227 |   .invoke_handler(tauri::generate_handler![   <-- the macro-generated dispatch await

warning: large future with a size of 31976 bytes  --> install.rs:643:1   install_fabric_version
warning: large future with a size of 32040 bytes  --> install.rs:1088:1  install_forge_version
warning: large future with a size of 31464 bytes  --> install.rs:806:1   install_quilt_version
```

A few notes:

- `install_modpack` is the command that produced the crash from the issue. The lint sees 73,032 bytes for it here; the incident analysis measured the raw state machine at 112,592 bytes via `-Zprint-type-sizes` on a release build — same code, different observation point and layout, so consistent enough.
- `ws_connect`, the ~329 KB case mentioned in the issue, only grew past the threshold in commits after the one tested here, so it rightly stays quiet in this run.
- Since the A/B runs share toolchain, flags and crate, the delta is exactly this PR: stock clippy never sees any of the four command futures.

So from the reporter's side this fully covers the real-world case — the diagnostics land on user-editable code, and the suggested `Box::pin` is literally the fix we shipped in the app after the incident.

One small thing I noticed while checking the branch: the two boundary shapes from your test matrix — the external-crate-macro future (stays silent) and the span-preserving proc macro (`with_span!`, fires) — don't seem to be covered by the committed tests yet; `tests/ui/large_futures.rs` only carries the `macro_rules!` cases. Since that silence boundary is the main behavioral line this PR draws, I'd be happy to add those two cases to the branch if you'd like.

**Comment by 0xull** [comment]
hi @dIT8Zv this is great, thank you for running it against the real app, that helps answer the tauri span question completely.

and yeah, please, i'd actually like to add those two boundary cases myself so they're in the test suite rather than just described in the pr, since the silent external! case is the main thing proving the change didn't over-broaden. i'll push them to the branch and tag you. thanks again for the thorough A/B, that's genuinely more than i expected.

**Comment by blyxyas** [comment]
Hello @dIT8Zv, it seems that you've used an AI/LLM for both the original opening and discussion of this issue. Please note that we have an [LLM usage policy](https://forge.rust-lang.org/policies/llm-usage.html), and generating issues or issue comments is banned.

@0xull, please proceed with caution when implementing your fix. Information generated by LLMs is not trustworthy, and you should always double-check everything by yourself.

**Comment by 0xull** [comment]
hi @blyxyas, thank you for the reply. i built and tested the expected changes locally and confirming the actual output aligns, as well as with the rest of the ui tests. so far, for the open questions i mentioned in the pr, i'm hoping the maintainers point me in the right direction if my assumptions are wrong. thanks again.

hi @dIT8Zv , i have added the rest of the test cases along to the commit. happy to walk through any of it.
